<a href="https://colab.research.google.com/github/Estteves/Estteves-Portfolio/blob/main/An%C3%A1lise_Explorat%C3%B3ria_de_Dados_de_Log%C3%ADstica.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img src="https://raw.githubusercontent.com/andre-marcos-perez/ebac-course-utils/main/media/logo/newebac_logo_black_half.png" alt="ebac-logo">

---

# **Módulo** | Análise de Dados: Análise Exploratória de Dados de Logística I
Caderno de **Exercícios**<br>
Professor [André Perez](https://www.linkedin.com/in/andremarcosperez/)

---

# **Tópicos**

<ol type="1">
  <li>Introdução ao Kaggle;</li>
  <li>Introdução ao problema de negócios;</li>
  <li>Exploração de dados.</li>
</ol>


---

# **Exercícios**

Este *notebook* deve servir como um guia para a construção da sua própria análise exploratória de dados. Fique a vontate para copiar os códigos da aula mas busque explorar os dados ao máximo. Por fim, publique seu *notebook* no [Kaggle](https://www.kaggle.com/).

---

# **Análise Exploratória de Dados de Logística**

## 1\. Contexto


O Loggi Benchmark for Urban Delivery contém conjuntos de dados e scripts de benchmark para problemas de grande escala usando como teste algumas das maiores cidades do Brasil. O conjunto de dados simula o desafio de uma grande empresa de entrega na última milha da cadeia de suprimentos. Em todo o mundo, áreas metropolitanas como o Rio de Janeiro testemunham a entrega de dezenas ou mesmo centenas de milhares de encomendas e correspondências. As entregas nessas áreas são inesperadamente distribuídas de forma muito desigual. Áreas ricas e densamente povoadas conduzem a maior parte das entregas. Além dos fatores econômicos, os dados também consideram a geografia dos locais representados e as distâncias reais de deslocamento nas ruas.
O objetivo é realizar uma análise exploratória dos dados utilizando as bibliotecas Python json e pandas.

## 2\. Pacotes e bibliotecas

In [ ]:
# importe todas as suas bibliotecas aqui, siga os padrões do PEP8:
import json
import pandas as pd

## 3\. Exploração de dados

In [ ]:
# faça o código de exploração de dados:
#Coletar e carregar dados
!wget -q "https://raw.githubusercontent.com/andre-marcos-perez/ebac-course-utils/main/dataset/deliveries.json" -O deliveries.json

with open('deliveries.json', mode='r', encoding='utf8') as file:
  data = json.load(file)

In [ ]:
# - wrangling da estrutura;
deliveries_df = pd.DataFrame(data)
deliveries_df.head()

name region                                             origin  \
0  cvrp-2-df-33   df-2  {'lng': -48.05498915846707, 'lat': -15.8381445...   
1  cvrp-2-df-73   df-2  {'lng': -48.05498915846707, 'lat': -15.8381445...   
2  cvrp-2-df-20   df-2  {'lng': -48.05498915846707, 'lat': -15.8381445...   
3  cvrp-1-df-71   df-1  {'lng': -47.89366206897872, 'lat': -15.8051175...   
4  cvrp-2-df-87   df-2  {'lng': -48.05498915846707, 'lat': -15.8381445...   

   vehicle_capacity                                         deliveries  
0               180  [{'id': '313483a19d2f8d65cd5024c8d215cfbd', 'p...  
1               180  [{'id': 'bf3fc630b1c29601a4caf1bdd474b85', 'po...  
2               180  [{'id': 'b30f1145a2ba4e0b9ac0162b68d045c3', 'p...  
3               180  [{'id': 'be3ed547394196c12c7c27c89ac74ed6', 'p...  
4               180  [{'id': 'a6328fb4dc0654eb28a996a270b0f6e4', 'p...

a coluna origin contem dados nested ou aninhados na estrutura do JSON. Vamos normalizar a coluna com uma operação conhecida como flatten ou achatamento que transforma cada chave do JSON em uma nova coluna.

In [ ]:
hub_origin_df = pd.json_normalize(deliveries_df["origin"])
hub_origin_df.head()

lng        lat
0 -48.054989 -15.838145
1 -48.054989 -15.838145
2 -48.054989 -15.838145
3 -47.893662 -15.805118
4 -48.054989 -15.838145

Com o dados achatados, vamos junta-los ao conjunto de dados principal:

In [ ]:
deliveries_df = pd.merge(left=deliveries_df, right=hub_origin_df, how='inner', left_index=True, right_index=True)
deliveries_df.head()

name region                                             origin  \
0  cvrp-2-df-33   df-2  {'lng': -48.05498915846707, 'lat': -15.8381445...   
1  cvrp-2-df-73   df-2  {'lng': -48.05498915846707, 'lat': -15.8381445...   
2  cvrp-2-df-20   df-2  {'lng': -48.05498915846707, 'lat': -15.8381445...   
3  cvrp-1-df-71   df-1  {'lng': -47.89366206897872, 'lat': -15.8051175...   
4  cvrp-2-df-87   df-2  {'lng': -48.05498915846707, 'lat': -15.8381445...   

   vehicle_capacity                                         deliveries  \
0               180  [{'id': '313483a19d2f8d65cd5024c8d215cfbd', 'p...   
1               180  [{'id': 'bf3fc630b1c29601a4caf1bdd474b85', 'po...   
2               180  [{'id': 'b30f1145a2ba4e0b9ac0162b68d045c3', 'p...   
3               180  [{'id': 'be3ed547394196c12c7c27c89ac74ed6', 'p...   
4               180  [{'id': 'a6328fb4dc0654eb28a996a270b0f6e4', 'p...   

         lng        lat  
0 -48.054989 -15.838145  
1 -48.054989 -15.838145  
2 -48.054989 -15.838145  
3 -47.893662 -15.805118  
4 -48.054989 -15.838145

Após separar os dados da coluna 'origin' vamos removê-la do Dataframe 'deliveries_df' e alterar a posição das colunas 'lng' e 'lat'

In [ ]:
deliveries_df = deliveries_df.drop("origin", axis=1)
deliveries_df = deliveries_df[["name", "region", "lng", "lat", "vehicle_capacity", "deliveries"]]
deliveries_df.head()

name region        lng        lat  vehicle_capacity  \
0  cvrp-2-df-33   df-2 -48.054989 -15.838145               180   
1  cvrp-2-df-73   df-2 -48.054989 -15.838145               180   
2  cvrp-2-df-20   df-2 -48.054989 -15.838145               180   
3  cvrp-1-df-71   df-1 -47.893662 -15.805118               180   
4  cvrp-2-df-87   df-2 -48.054989 -15.838145               180   

                                          deliveries  
0  [{'id': '313483a19d2f8d65cd5024c8d215cfbd', 'p...  
1  [{'id': 'bf3fc630b1c29601a4caf1bdd474b85', 'po...  
2  [{'id': 'b30f1145a2ba4e0b9ac0162b68d045c3', 'p...  
3  [{'id': 'be3ed547394196c12c7c27c89ac74ed6', 'p...  
4  [{'id': 'a6328fb4dc0654eb28a996a270b0f6e4', 'p...

Agora, vamos renomear as colunas 'lng' e 'lat'

In [ ]:
deliveries_df.rename(columns={"lng": "hub_lng", "lat": "hub_lat"}, inplace=True)
deliveries_df.head()

name region    hub_lng    hub_lat  vehicle_capacity  \
0  cvrp-2-df-33   df-2 -48.054989 -15.838145               180   
1  cvrp-2-df-73   df-2 -48.054989 -15.838145               180   
2  cvrp-2-df-20   df-2 -48.054989 -15.838145               180   
3  cvrp-1-df-71   df-1 -47.893662 -15.805118               180   
4  cvrp-2-df-87   df-2 -48.054989 -15.838145               180   

                                          deliveries  
0  [{'id': '313483a19d2f8d65cd5024c8d215cfbd', 'p...  
1  [{'id': 'bf3fc630b1c29601a4caf1bdd474b85', 'po...  
2  [{'id': 'b30f1145a2ba4e0b9ac0162b68d045c3', 'p...  
3  [{'id': 'be3ed547394196c12c7c27c89ac74ed6', 'p...  
4  [{'id': 'a6328fb4dc0654eb28a996a270b0f6e4', 'p...

A coluna deliveries contem dados uma lista de dados nested ou aninhados na estrutura do JSON. Vamos normalizar a coluna com uma operação conhecida como explode ou explosão que transforma cada elemento da lista em uma linha. Por fim, faremos os flatten ou achatamento do resultado coluna:

In [ ]:
deliveries_exploded_df = deliveries_df[["deliveries"]].explode("deliveries")
deliveries_exploded_df.head()

deliveries
0  {'id': '313483a19d2f8d65cd5024c8d215cfbd', 'po...
0  {'id': '320c94b17aa685c939b3f3244c3099de', 'po...
0  {'id': '3663b42f4b8decb33059febaba46d5c8', 'po...
0  {'id': 'e11ab58363c38d6abc90d5fba87b7d7', 'poi...
0  {'id': '54cb45b7bbbd4e34e7150900f92d7f4b', 'po...

In [ ]:
deliveries_normalized_df = pd.concat([
  pd.DataFrame(deliveries_exploded_df["deliveries"].apply(lambda record: record["size"])).rename(columns={"deliveries": "delivery_size"}),
  pd.DataFrame(deliveries_exploded_df["deliveries"].apply(lambda record: record["point"]["lng"])).rename(columns={"deliveries": "delivery_lng"}),
  pd.DataFrame(deliveries_exploded_df["deliveries"].apply(lambda record: record["point"]["lat"])).rename(columns={"deliveries": "delivery_lat"}),
], axis= 1)
deliveries_normalized_df.head()

delivery_size  delivery_lng  delivery_lat
0              9    -48.116189    -15.848929
0              2    -48.118195    -15.850772
0              1    -48.112483    -15.847871
0              2    -48.118023    -15.846471
0              7    -48.114898    -15.858055

Após separar os dados da coluna 'deliveries' vamos removê-la do Dataframe 'deliveries_df' e concatenar a direita os dados do 'deliveries_normalized_df' com o 'deliveries_df'

In [ ]:
deliveries_df = deliveries_df.drop("deliveries", axis=1)
deliveries_df = pd.merge(left=deliveries_df, right=deliveries_normalized_df, how='right', left_index=True, right_index=True)
deliveries_df.reset_index(inplace=True, drop=True)
deliveries_df.head()

name region    hub_lng    hub_lat  vehicle_capacity  delivery_size  \
0  cvrp-2-df-33   df-2 -48.054989 -15.838145               180              9   
1  cvrp-2-df-33   df-2 -48.054989 -15.838145               180              2   
2  cvrp-2-df-33   df-2 -48.054989 -15.838145               180              1   
3  cvrp-2-df-33   df-2 -48.054989 -15.838145               180              2   
4  cvrp-2-df-33   df-2 -48.054989 -15.838145               180              7   

   delivery_lng  delivery_lat  
0    -48.116189    -15.848929  
1    -48.118195    -15.850772  
2    -48.112483    -15.847871  
3    -48.118023    -15.846471  
4    -48.114898    -15.858055

Informações sobre o DataFrame deliveries_df

In [ ]:
deliveries_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636149 entries, 0 to 636148
Data columns (total 8 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   name              636149 non-null  object 
 1   region            636149 non-null  object 
 2   hub_lng           636149 non-null  float64
 3   hub_lat           636149 non-null  float64
 4   vehicle_capacity  636149 non-null  int64  
 5   delivery_size     636149 non-null  int64  
 6   delivery_lng      636149 non-null  float64
 7   delivery_lat      636149 non-null  float64
dtypes: float64(4), int64(2), object(2)
memory usage: 38.8+ MB


In [ ]:
# - exploração do schema;
# Visualizando as 5 primeiras linhas do DataFrame
deliveries_df.head(n=5)

name region    hub_lng    hub_lat  vehicle_capacity  delivery_size  \
0  cvrp-2-df-33   df-2 -48.054989 -15.838145               180              9   
1  cvrp-2-df-33   df-2 -48.054989 -15.838145               180              2   
2  cvrp-2-df-33   df-2 -48.054989 -15.838145               180              1   
3  cvrp-2-df-33   df-2 -48.054989 -15.838145               180              2   
4  cvrp-2-df-33   df-2 -48.054989 -15.838145               180              7   

   delivery_lng  delivery_lat  
0    -48.116189    -15.848929  
1    -48.118195    -15.850772  
2    -48.112483    -15.847871  
3    -48.118023    -15.846471  
4    -48.114898    -15.858055

Visualizando as colunas e seus respectivos tipos de dados.

In [ ]:
deliveries_df.dtypes

name                 object
region               object
hub_lng             float64
hub_lat             float64
vehicle_capacity      int64
delivery_size         int64
delivery_lng        float64
delivery_lat        float64
dtype: object

Visualizando apenas atributos categóricos

In [ ]:
deliveries_df.select_dtypes("object").describe().transpose()

count unique           top    freq
name    636149    199  cvrp-1-df-87    5636
region  636149      3          df-1  304708

#Insight 1
Podemos observar que o nome que mais aparece é o 'cvrp-1-df-87', aparece 5.636 vezes no DataFrame. Além disso, temos 199 nomes únicos.

Podemos observar também que a região que mais temos entregas é a 'df-1', pois consta 304.708 vezes no DataFrame.

Visualizando apenas atributos numéricos. (Excluindo do DataFrame os dados categóricos 'name' e 'region')

In [ ]:
deliveries_df.drop(["name", "region"], axis=1).select_dtypes('int64').describe().transpose()

count        mean       std    min    25%    50%    75%  \
vehicle_capacity  636149.0  180.000000  0.000000  180.0  180.0  180.0  180.0   
delivery_size     636149.0    5.512111  2.874557    1.0    3.0    6.0    8.0   

                    max  
vehicle_capacity  180.0  
delivery_size      10.0

#Insight 2
Podemos observar que não existe variação na capacidade dos veículos de entrega, ou seja, todos possuem capacidade de 180.

Podemos observar também que o tamanho da carga da entrega varia entre 1 e 10, o tamanho médio é de 5.5, com variação de aproximadamente 2.87 para mais ou para menos.

In [ ]:
#Verificando se alguma coluna possui dado faltante
deliveries_df.isna().any()

name                False
region              False
hub_lng             False
hub_lat             False
vehicle_capacity    False
delivery_size       False
delivery_lng        False
delivery_lat        False
dtype: bool

Todas as colunas estão preenchidas.